### ISSUE TITLE
Unable to deserialize enum from [FromQuery] with (or without) JsonStringEnumMemberConverter

### ISSUE THREAD

**Issue by diegosasw** [Macross-Software/core#30]
I cannot use `JsonStringEnumMemberConverter` with `[FromQuery]` in my .NET 6 AspNetCore controllers to deserialize enum values matching a specific `EnumMember`. It seems the binder fails (with and without JsonStringEnumMemberConverter)

See my next comment for a repo to reproduce and the different approaches taken.

**Comment by diegosasw** [comment]
I have added a repo to reproduce it.

1. Here is with the explicit converter on the enum itself: https://gitlab.com/sunnyatticsoftware/issues/string-to-enum-mvc/-/tree/feature/2-macross-json-explicit-converter

```
[JsonConverter(typeof(JsonStringEnumMemberConverter))]
public enum Status
{
    [EnumMember(Value = "open")]
    Open,
    
    [EnumMember(Value = "on-hold")]
    OnHold
}
```

2. Here is with the json serializer options converter registered: https://gitlab.com/sunnyatticsoftware/issues/string-to-enum-mvc/-/tree/feature/3-macross-json-implicit-converter
```
services
    .AddOpenApi()
    .AddControllers()
    .AddJsonOptions(
        options =>
        {
            options.JsonSerializerOptions.Converters.Clear(); // this line won't make any difference
            options.JsonSerializerOptions.Converters.Add(new JsonStringEnumMemberConverter());
        });
```

In both cases, a request such as `http://localhost:5000/Samples?Foo=bar&Status=on-hold` fails to deserialize the query string into `QueryModel` 
```
public class QueryModel
{
    /// <summary>
    /// The foo parameter
    /// </summary>
    /// <example>bar</example>
    public string? Foo { get; init; } = null;
    
    /// <summary>
    /// The status
    /// </summary>
    /// <example>on-hold</example>
    public Status? Status { get; set; } = null;
}

public enum Status
{
    [EnumMember(Value = "open")]
    Open,
    
    [EnumMember(Value = "on-hold")]
    OnHold
}
```
but the `http://localhost:5000/Samples?Foo=bar&Status=open` is successful, as it can be viewed by running the sample and interacting with the swagger UI
1. Clone the repository
2. Run with `dotnet run --project ./src/Sample.WebApi/Sample.WebApi.csproj`
3. Open `http://localhost:5000/swagger` and send http requests through the UI or send explicit http get requests like `http://localhost:5000/Samples?Foo=bar&Status=on-hold`

**Comment by CodeBlanch** [comment]
Thanks @diegosasw! My guess would be there is a quirk to what AspNetCore does with the `FromQuery` attribute but let me look into it and I'll report back my findings.

**Comment by CodeBlanch** [comment]
@diegosasw

So it looks like the entire `System.Text.Json` engine is invoked as an [IInputFormatter](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.formatters.iinputformatter?view=aspnetcore-6.0) which only fires on the body.

There are some details on [Model Binding in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/mvc/models/model-binding?view=aspnetcore-6.0#frombody-attribute).

What seems to be firing for `FromQuery` in your repo is [ComplexTypeModelBinder](https://github.com/dotnet/aspnetcore/blob/main/src/Mvc/Mvc.Core/src/ModelBinding/Binders/ComplexObjectModelBinder.cs). First time I've seen that, doesn't appear to have any extension points.

What this means is the answer is essentially: In ASP.NET Core JSON isn't supported in query strings, only bodies.

BUT I did manage to get it working. FYI I have never authored an `IModelBinder` implementation before so I don't know what I don't know. Use this at your own risk 😄

```csharp
public class QueryModel
{
    public string? Foo { get; init; } = null;

    [ModelBinder(typeof(JsonStringEnumMemberModelBinder<Status>))]
    public Status? Status { get; set; } = null;
}

internal static class JsonStringEnumMemberModelBinderHelper
{
    internal static JsonStringEnumMemberConverter Factory { get; } = new JsonStringEnumMemberConverter();
    internal static JsonSerializerOptions Options { get; } = new();
}

public class JsonStringEnumMemberModelBinder<T> : IModelBinder
    where T : struct, Enum
{
    private static readonly JsonConverter<T> _Converter
        = (JsonConverter<T>)JsonStringEnumMemberModelBinderHelper.Factory.CreateConverter(typeof(T), JsonStringEnumMemberModelBinderHelper.Options);

    public Task BindModelAsync(ModelBindingContext bindingContext)
    {
        string? value = bindingContext.ValueProvider
            .GetValue(bindingContext.FieldName)
            .FirstValue;

        if (!string.IsNullOrEmpty(value))
        {
            Span<byte> data = stackalloc byte[256];
            data[0] = (byte)'\"';
            int numberOfBytes = Encoding.UTF8.GetBytes(value, data.Slice(1, 254));
            if (numberOfBytes > 0)
            {
                data[1 + numberOfBytes] = (byte)'\"';

                Utf8JsonReader reader = new(data.Slice(0, numberOfBytes + 2));
                if (reader.Read())
                {
                    T enumValue = _Converter.Read(ref reader, typeof(T), JsonStringEnumMemberModelBinderHelper.Options);

                    bindingContext.Result = ModelBindingResult.Success(enumValue);
                }
            }
        }

        return Task.CompletedTask;
    }
}
```

Basically what I did there was create a model binder which will jump into `System.Text.Json` + `JsonStringEnumMemberConverter`.
